### PR TITLE
A basic noise analysis script seems to work.

### PR DESCRIPTION
**Pull request by joefowler**
Well, I'm not sure this is what you have in mind, but I reached a stopping point and think it's worth considering for merge now.

### CODE REVIEW

**Comment by joefowler** [comment]
Oops. It's failing tests because I didn't require `ARMA`. Do you remember how to do that for packages that aren't in `METADATA`? I think you figured this out once.

**Comment by ggggggggg** [comment]
Look at deps/build.jl, it already requires one package that isn't in METADATA so you can copy that.

File: test/noise_analysis.jl [hunk 1]
```diff
@@ -0,0 +1,58 @@
+using Base.Test
+using ARMA
+using HDF5
+using Pope: NoiseAnalysis
+
+function compare(m1::ARMAModel, m2::ARMAModel)
```
Comment (ggggggggg) [inline test/noise_analysis.jl]: Why not define  [`Base.isapprox`](https://docs.julialang.org/en/stable/stdlib/math/#Base.isapprox)

File: scripts/noise_analysis.jl [hunk 2]
```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env julia
+using ArgParse
+using HDF5
+using ARMA
+using Pope: NoiseAnalysis, LJH
+
+function analyze_one_file(filename::AbstractString, channum::Integer,
```
Comment (ggggggggg) [inline scripts/noise_analysis.jl]: This is LJHFile specific, but I'm not going to let that hold things up. We can easily change it later.

File: src/NoiseAnalysis.jl [hunk 3]
```diff
@@ -0,0 +1,231 @@
+"""
```
Comment (ggggggggg) [inline src/NoiseAnalysis.jl]: Docstrings on Modules don't appear anywhere in the doc webpage as far as I can tell. This should go in a docs/src/XX.md file.

File: src/NoiseAnalysis.jl [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        end
+        g = h5file[chanstring]["noise"]
+        return hdf5load(g)
+    end
+end
+
+function hdf5load(g::HDF5.DataFile)
+    acorr = g["autocorr"][:]
+    psd = g["powerspectrum"][:]
+    sampused = read(attrs(g)["samplesused"])
```
Comment (ggggggggg) [inline src/NoiseAnalysis.jl]: I don't think we should use any attrs.

File: src/NoiseAnalysis.jl [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+"""
+    round_up_dft_length(n)
+
+A length at least as large as `n` convenient for a DFT: (1, 3, or 5) x any power
+of 2. This ensures that we never do DFTs on an especially bad vector size.
+"""
+function round_up_dft_length(n::Integer)
+    pow2 = 2^ceil(log2(n))
+    if n > 0.75*pow2
+        return Int(round(pow2))
```
Comment (ggggggggg) [inline src/NoiseAnalysis.jl]: I think `round(Int,pow2)` is preferred.

File: src/NoiseAnalysis.jl [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    ac ./ (m:-1:m-nlags+1)
+end
+
+# Power spectra and window functions:
+"""
+    bartlett(n)
+
+Bartlett window of length `n`.
+"""
+function bartlett(n::Integer)
```
Comment (ggggggggg) [inline src/NoiseAnalysis.jl]: Do you define these to avoid depending on `DSP`?

File: src/NoiseAnalysis.jl [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    nsamp = 2(nfreq-1)
+    if length(data) < nsamp
+        error("data must be at least 2(nfreq-1) in length.")
+    end
+
+    # If you have any appreciable power at DC and wish to understand the low-freq
+    # power spectrum, you want the hann window or no window, because only these
+    # have zero leakage for bins 3:end (zero leakage into bin 2, also, if using
+    # no window at all).
+    window = hann(nsamp)
```
Comment (ggggggggg) [inline src/NoiseAnalysis.jl]: The window should be an argument? eg `compute_psd(data::AbstractVector, nfreq::Integer, dt::Real; max_exc=1e99, windowfunc=hann)`, `window=windowfunc(nsamp)`.

File: src/NoiseAnalysis.jl [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    window = window / sqrt(sum(window.^2)) # proper normalization
+
+    # Break data up into half-overlapping segments of the right separation
+    nseg = Int(ceil(length(data)/nsamp))
```
Comment (ggggggggg) [inline src/NoiseAnalysis.jl]: `ceil(Int,length(data/nsamp)`

File: test/noise_analysis.jl [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        fname2 = tempname()*".hdf5"
+        h5open(fname2, "w") do f
+            g1 = g_create(f, "top")
+            g2 = g_create(g1, "mid")
+            g3 = g_create(g2, "low")
+            NoiseAnalysis.hdf5save(g3, noiseresult)
+        end
+        h5open(fname2, "r") do f
+            nr2 = NoiseAnalysis.hdf5load(f["top/mid/low"])
+            compare(noiseresult, nr2)
```
Comment (ggggggggg) [inline test/noise_analysis.jl]: If you define `isapprox` this can be `@test noiseresult ≈ nr2`

File: test/noise_analysis.jl [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        noiseresult = NoiseAnalysis.NoiseResult(acorr, psd, used, freqstep, source, model)
+
+        fname1 = tempname()*".hdf5"
+        channum = rand(1:99)
+        NoiseAnalysis.hdf5save(fname1, channum, noiseresult)
+        nr1 = NoiseAnalysis.hdf5load(fname1, channum)
+        compare(noiseresult, nr1)
+
+        fname2 = tempname()*".hdf5"
+        h5open(fname2, "w") do f
```
Comment (ggggggggg) [inline test/noise_analysis.jl]: Doesn't one of your load functions require a integer input? Lets use a file structure like that in at least one test.

**Comment by joefowler** [comment]
Thank you for the tip on deps/build.jl. I remembered you showing me that workaround, but I thought it would be in REQUIRE.

I agree that we should merge soon. Let me make a few changes based on your suggestions, push them, see if the tests pass, and maybe we can merge this afternoon.

File: test/noise_analysis.jl (hunk 10 above)
Comment (joefowler) [reply test/noise_analysis.jl]: That's in the earlier test, the one with `fname1`.

**Comment by joefowler** [comment]
Tests are failing because `NLopt` and `HDF5` packages won't build. I have no idea how to fix this problem. You?

**Comment by ggggggggg** [comment]
I'm not sure whats going on with `HDF5` since it [works on other automated builds,](https://travis-ci.org/usnistgov/Pope.jl/builds/319427515#L723) but for `NLopt` I think you will need to add any apt packages to the `.travis.yml` file under "addons: apt: packages:".

**Comment by joefowler** [comment]
You mean "apt" as in Linux packages? How in the world do I figure out what apt packages I even need?

**Comment by ggggggggg** [comment]
https://github.com/JuliaOpt/NLopt.jl/blob/master/deps/build.jl#L8

 or 

https://travis-ci.org/usnistgov/Pope.jl/builds/320319829#L769

**Comment by joefowler** [comment]
Thanks for the tips. You can see on [ARMA.jl](https://travis-ci.org/joefowler/ARMA.jl/) that it was also failing for similar reasons, but your pointers helped me to fix it. I'll do the same to the Pope config of Travis.

**Comment by ggggggggg** [comment]
tests pass, merge when you think it's ready

**Pull request merged**